### PR TITLE
Document batch support

### PR DESCRIPTION
**Pull request by MarinPostma**
This pr adds support for document batches in milli. It changes the API of the `IndexDocuments` builder by adding a `add_documents` method. The API of the updates is changed a little, with the `UpdateBuilder` being renamed to `IndexerConfig` and being passed to the update builders. This makes it easier to pass around structs that need to access the indexer config, rather that extracting the fields each time. This change impacts many function signatures and simplify them.

The change in not thorough, and may require another PR to propagate to the whole codebase. I restricted to the necessary for this PR.

### CODE REVIEW

File: milli/src/update/index_documents/mod.rs [hunk 1]
```diff
@@ -55,120 +51,115 @@ pub enum IndexDocumentsMethod {
...
-    pub fn execute<R, F>(
-        self,
-        reader: DocumentBatchReader<R>,
-        progress_callback: F,
-    ) -> Result<DocumentAdditionResult>
+    /// Adds a batch of documents to the current builder.
+    ///
+    /// Since the documents are progressively added to the writer, a failure will cause a stale
+    /// builder, and the builder must be discarded.
+    #[must_use = "a builder must not be used after a add document operation reported an error."]
```
Comment (Kerollmops) [inline milli/src/update/index_documents/mod.rs:121]: It is not how `must_use` works, [here it is advertising the user that he is not using the u64 output by this method](https://rust-lang.github.io/rfcs/1940-must-use-functions.html).

Comment (Kerollmops) [reply milli/src/update/index_documents/mod.rs:121]: Can you add a point about what this return `u64` is, please?

File: milli/src/update/index_documents/transform.rs [hunk 2]
```diff
@@ -84,56 +81,73 @@ fn find_primary_key(index: &DocumentsBatchIndex) -> Option<&str> {
         .map(String::as_str)
 }
 
-impl Transform<'_, '_> {
+impl<'i, 'a> Transform<'a, 'i> {
```
Comment (Kerollmops) [inline milli/src/update/index_documents/transform.rs:84]:
```suggestion
impl<'a, 'i> Transform<'a, 'i> {
```

File: milli/src/update/index_documents/transform.rs [hunk 3]
```diff
@@ -230,38 +244,40 @@ impl Transform<'_, '_> {
             documents_seen: documents_count,
         });
 
+        self.index.put_fields_ids_map(wtxn, &fields_ids_map)?;
+        self.index.put_primary_key(wtxn, &primary_key_name)?;
```
Comment (Kerollmops) [inline milli/src/update/index_documents/transform.rs:248]: Ok, now the transform struct writes into the database, to keep a constant state? For example the new fields it found when iterating through documents?

File: milli/src/update/settings.rs [hunk 4]
```diff
@@ -118,14 +110,10 @@ impl<'a, 't, 'u, 'i> Settings<'a, 't, 'u, 'i> {
...
         }
     }
 
-    pub fn log_every_n(&mut self, n: usize) {
-        self.log_every_n = Some(n);
-    }
-
     pub fn reset_searchable_fields(&mut self) {
         self.searchable_fields = Setting::Reset;
     }
```
Comment (Kerollmops) [inline milli/src/update/settings.rs:119]: Why don't you remove those functions too?

File: milli/src/update/update_builder.rs [hunk 5]
```diff
@@ -1,130 +0,0 @@
-use grenad::CompressionType;
```
Comment (Kerollmops) [inline milli/src/update/update_builder.rs:1]: Have you deleted this file?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: milli/src/update/update_builder.rs (hunk 5 above)
Comment (MarinPostma) [reply milli/src/update/update_builder.rs:1]: it was renamed to `indexer_config.rs`, but git hasn't picked it up

File: milli/src/update/settings.rs (hunk 4 above)
Comment (MarinPostma) [reply milli/src/update/settings.rs:119]: they are not part of the `IndexerConfig`, we could create another struct to hold them, and have it passed to the `Settings` constructor, like we do for the `IndexDocumentsConfig`, but that seemed out of scope for this PR.

File: milli/src/update/index_documents/transform.rs (hunk 3 above)
Comment (MarinPostma) [reply milli/src/update/index_documents/transform.rs:248]: that's right

File: milli/src/update/settings.rs (hunk 4 above)
Comment (Kerollmops) [reply milli/src/update/settings.rs:119]: Good point!

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Looks good to me, you should do a rustfmt pass :)

**Review by Kerollmops — DISMISSED** [review]

**Comment by MarinPostma** [comment]
bors try

**Comment by Kerollmops** [comment]
Ok, do we want to merge that now? I would prefer that we wait for the fix of @irevoire about https://github.com/meilisearch/MeiliSearch/issues/2082 first.

**Comment by MarinPostma** [comment]
@Kerollmops, I think we can merge that now

**Comment by Kerollmops** [comment]
Yup, indeed!
bors merge

**Comment by MarinPostma** [comment]
bors try

**Comment by Kerollmops** [comment]
bors merge

**Review by Kerollmops — APPROVED** [review]
My bad 🤦‍♂️
bors merge

**Pull request merged**